qiskit/aqua/operators/primitive_ops/pauli_sum_op.py

```diff
@@ -0,0 +1,377 @@
+from ..operator_base import OperatorBase
+from .primitive_op import PrimitiveOp
+
+logger = logging.getLogger(__name__)
+
+
+class PauliSumOp(PrimitiveOp):
+    """Class for Operators backend by Terra's ``SparsePauliOp`` class."""
+
+    def __init__(
+            self,
+            primitive: SparsePauliOp,
+            coeff: Union[int, float, complex, ParameterExpression] = 1.0,
+    ) -> None:
+        """
+        Args:
+            primitive: The SparsePauliOp which defines the behavior of the underlying function.
+            coeff: A coefficient multiplying the primitive.
+
+        Raises:
+            TypeError: invalid parameters.
+        """
+        if not isinstance(primitive, SparsePauliOp):
+            raise TypeError(
+                f"PauliSumOp can only be instantiated with SparsePauliOp, not {type(primitive)}"
+            )
+
+        super().__init__(primitive, coeff=coeff)
+
+    def primitive_strings(self) -> Set[str]:
+        return {"SparsePauliOp"}
+
+    @property
+    def num_qubits(self) -> int:
+        return self.primitive.num_qubits  # type: ignore
+
+    def add(self, other: OperatorBase) -> OperatorBase:
+        if not self.num_qubits == other.num_qubits:
+            raise ValueError(
+                f"Sum of operators with different numbers of qubits, {self.num_qubits} and "
+                f"{other.num_qubits}, is not well defined"
+            )
+
+        if isinstance(other, PauliSumOp):
+            return PauliSumOp(
+                self.coeff * self.primitive + other.coeff * other.primitive, coeff=1  # type: ignore
+            )
+
+        from .pauli_op import PauliOp
+
+        if isinstance(other, PauliOp):
+            return PauliSumOp(
+                self.coeff * self.primitive  # type: ignore
+                + other.coeff * SparsePauliOp(other.primitive)
+            )
+
+        return SummedOp([self, other])
+
+    def mul(self, scalar: Union[int, float, complex, ParameterExpression]) -> OperatorBase:
+        if isinstance(scalar, (int, float, complex)) and scalar != 0:
+            return PauliSumOp(scalar * self.primitive, coeff=self.coeff)  # type: ignore
+
+        return super().mul(scalar)
+
+    def adjoint(self) -> OperatorBase:
+        return PauliSumOp(
+            self.primitive.conjugate(), coeff=self.coeff.conjugate()  # type:ignore
+        )
+
+    def equals(self, other: OperatorBase) -> bool:
+        self_reduced, other_reduced = self.reduce(), other.reduce()
+        if not isinstance(other_reduced, PauliSumOp):
+            return False
+
+        if isinstance(self_reduced.coeff, ParameterExpression) or isinstance(
+                other_reduced.coeff, ParameterExpression
+        ):
+            return (
+                self_reduced.coeff == other_reduced.coeff
+                and self_reduced.primitive == other_reduced.primitive  # type:ignore
+            )
+        return (
+            len(self_reduced) == len(other_reduced)
```

Comment:
If this line is here for efficiency, it's probably not necessary. Note that in `SparsePauliOp` a check for length is not made when checking for equality. This is probably because the equality check calls `np.allclose`, and that function likely does a size check first.

Reply:
This is necessary for broadcasting of np.allclose.

e.g. The code
```python
((X^X) +(Y^X)).primitive == ((X^X) +(Y^X) + (Y^Y)).primitive
```
raises an error
```
ValueError: operands could not be broadcast together with shapes (2,) (3,) 
```

Reply:
You are correct. But, I misread the code in `SparsePauliOp`. The check for the same size is in a call to the parent `__eq__` method. Here:
https://github.com/Qiskit/qiskit-terra/blob/215bdf84e466fe5d55a05d74d461f53cfede226e/qiskit/quantum_info/operators/symplectic/sparse_pauli_op.py#L84
and here:
https://github.com/Qiskit/qiskit-terra/blob/215bdf84e466fe5d55a05d74d461f53cfede226e/qiskit/quantum_info/operators/base_operator.py#L76

So, unless I am mistaken (again) `PauliSumOp.equals` checks the dimensions twice.

Reply:
The check is only for dimension of operators (and num_qubits). We need to check the number of terms here.

e.g. The code
```python
((X^X) +(Y^X)).primitive == ((X^X) +(Y^X) + (Y^Y)).primitive
```
raises an error
```
ValueError: operands could not be broadcast together with shapes (2,) (3,) 
```

Reply:
I think we should fix this in upstream (Terra), and my code is just workaround.

Reply:
Agreed. This really looks like a bug in `SparsePauliOp`.